content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md

```diff
@@ -0,0 +1,75 @@
+
+# Implementation
+The implementation of the project builds on several ideas explored in class, such as peephole optimizations and intermediate representations.
+
+## Design
+The basic design of the project is to convert code which uses numpy into an intermediate representation, perform the peephole optimizations on the intermediate representation, and then execute the code.
+
+### Scope
+The project's proof of concept for "peephole improvements" is to implement code which does the `logsumexp` trick for people. In machine learning the `logsumexp` operation will come up somewhat frequently, such as when computing the log of a softmax (when coupled with another normalization term). However, it is easy for this operation to run into overflow issues when any value of $x$ is big.
+The logsumexp trick of replacing $\log\sum_{i=1}^n\exp(x_i)$ with $\max_{i\in[n]}(x_i) + \log\sum_{i=1}^n\exp(x_i - \max_{i\in[n]}(x_i))$
+with is a common way of dealing with this problem, since
+$$\log\sum_{i=1}^n\exp(x_i) = c + \log\sum_{i=1}^n\exp(x_i-c) \text{ for all } c,$$
+and so setting $c = \max_{i\in[n]}(x_i)$ guarantees that you only exponentiate terms
+$$x_i - c \leq 0,$$ avoiding overflow issues. 
+
+#### Logsumexp trick proof
+$$\begin{aligned}\log\sum_{i=1}^n\exp(x_i) &=\log\sum_{i=1}^n1\exp(x_i)\\\ &= \log\sum_{i=1}^n\exp(c-c)\exp(x_i)\\\ &= \log\left(\exp(c)\sum_{i=1}^n\exp(x_i-c)\right)\\\ &= c + \log\sum_{i=1}^n\exp(x_i-c) \end{aligned}$$
+
+### JAX and Jaxpr
+Thankfully, there is already a project which converts numpy code into an intermediate representation, then then executes it. [JAX](https://jax.readthedocs.io/en/latest/) is a project which combines autograd (automatic differentiation for python) and XLA to make it easier to write numpy code which runs on GPUs and lets you take derivatives automatically. This seemed like a better starting point for writing peephole improvements than trying to start from scratch because it already exists and has other useful features.
+
+Jaxpr is JAX's intermediate representation for numpy code. While obviously different from BRIL, I found it to be pretty similar and pretty simple, so I was happy to build on top of Jaxpr rather than trying to target some other intermediate representation such as XLA or LLVM. Jaxpr, similar to BRIL, organizes programs into closures with a list of lines of code made of (basically) `(operation, in_variables, out_variables, type/shape)` tuples.
+
+In order to use JAX, you need to import JAX's instrumented (among other things) version of `numpy` instead of using normal `numpy`. I've generally found this to be as easy as calling `import jax.numpy as np` and then forgetting about it.
+
+## Hardest parts
```

Comment:
Before jumping into the hardest parts, what do you think about an "Implementation" section to tell us more about exactly what you did? Presumably it's some kind of a pass? How does it recognize the pattern it's looking for? How flexible were you able to make the pattern-matching? That sort of thing.